File: src/services/ats/greenhouse.ts
import { logger } from "../../logger";
import { type AtsJobData, type Fetcher, greenhouseJobSchema } from "./types";

const log = logger.child({ component: "ats/greenhouse" });

export function parseGreenhouseUrl(url: string): { org: string; id: string } | null {
  try {
    const { hostname, pathname } = new URL(url);
    if (!hostname.includes("greenhouse.io")) return null;
    const segments = pathname.split("/").filter(Boolean);
    // boards.greenhouse.io/{org}/jobs/{id} and job-boards.greenhouse.io/{org}/jobs/{id}
    const jobsIdx = segments.indexOf("jobs");
    if (jobsIdx <= 0 || jobsIdx === segments.length - 1) return null;
    const org = segments[jobsIdx - 1];
    const id = segments[jobsIdx + 1];
    if (!org || !id) return null;
    return { org, id };
  } catch {
    return null;
  }
}

function extractCountry(officeLocation: string | null | undefined): string | null {
  if (!officeLocation) return null;
  // offices[].location is "City, Region, Country" — take the last comma-segment.
  const parts = officeLocation
    .split(",")
    .map((s) => s.trim())
    .filter(Boolean);
  return parts.length > 1 ? (parts[parts.length - 1] ?? null) : null;
}

export async function fetchGreenhouseJob(
  url: string,
  fetcher: Fetcher = fetch,
): Promise<AtsJobData | null> {
  const parsed = parseGreenhouseUrl(url);
  if (!parsed) return null;
  const { org, id } = parsed;

  const apiUrl = `https://boards-api.greenhouse.io/v1/boards/${org}/jobs/${id}`;

  let res: Response;
  try {
    res = await fetcher(apiUrl);
  } catch (err) {
    log.warn({ err, url }, "greenhouse fetch failed");
    return null;
  }

  if (!res.ok) {
    log.warn({ status: res.status, url }, "greenhouse non-ok response");
    return null;
  }

  let raw: unknown;
  try {
    raw = await res.json();
  } catch (err) {
    log.warn({ err, url }, "greenhouse json parse failed");
    return null;
  }

  const result = greenhouseJobSchema.safeParse(raw);
  if (!result.success) {
    log.warn({ url, issues: result.error.issues }, "greenhouse schema mismatch");
    return null;
  }

  const job = result.data;
  const primary = job.location?.name ?? "";
  const officeLocations = (job.offices ?? [])
    .map((o) => o.location)
    .filter((l): l is string => !!l);
  const locations =
    primary && !officeLocations.includes(primary) ? [primary, ...officeLocations] : officeLocations;

  return {
    source: "greenhouse",
    location: primary,
    locations,
    workplaceType: null,
    country: extractCountry(job.offices?.[0]?.location),
  };
}

File: job_finder/ats/greenhouse.py
from __future__ import annotations

from typing import ClassVar

from pydantic import BaseModel, ConfigDict, StrictInt

from job_finder.ats.models import AtsAvailable, unique_locations
from job_finder.urls import parse_http_url


class _GreenhouseWireModel(BaseModel):
    model_config: ClassVar[ConfigDict] = ConfigDict(strict=True)


class _GreenhouseLocation(_GreenhouseWireModel):
    name: str


class _GreenhouseOffice(_GreenhouseWireModel):
    id: StrictInt
    location: str | None = None


class _GreenhouseJob(_GreenhouseWireModel):
    id: StrictInt
    location: _GreenhouseLocation | None = None
    offices: list[_GreenhouseOffice] | None = None


def parse_greenhouse_url(url: str) -> tuple[str, str] | None:
    parsed = parse_http_url(url)
    if parsed is None or parsed.hostname is None:
        return None
    if parsed.hostname not in (
        "boards.greenhouse.io",
        "job-boards.greenhouse.io",
        "boards.eu.greenhouse.io",
    ):
        return None
    segments = tuple(segment for segment in parsed.path.split("/") if segment)
    try:
        jobs_index = segments.index("jobs")
    except ValueError:
        return None
    if jobs_index == 0 or jobs_index == len(segments) - 1:
        return None
    return segments[jobs_index - 1], segments[jobs_index + 1]


def parse_greenhouse_job(payload: object) -> AtsAvailable:
    job = _GreenhouseJob.model_validate(payload)
    primary = job.location.name if job.location else ""
    office_locations = tuple(office.location for office in job.offices or () if office.location)
    country = _country_from_location(job.offices[0].location) if job.offices else None
    return AtsAvailable(
        source="greenhouse",
        location=primary,
        locations=unique_locations(primary, office_locations),
        workplace_type=None,
        country=country,
    )


def _country_from_location(location: str | None) -> str | None:
    if not location:
        return None
    parts = tuple(part.strip() for part in location.split(",") if part.strip())
    return parts[-1] if len(parts) > 1 else None
